extmod/modbluetooth_nimble.c

```diff
@@ -0,0 +1,690 @@
+    ble_hs_cfg.gatts_register_cb = gatts_register_cb;
+    ble_hs_cfg.store_status_cb = ble_store_util_status_rr;
+
+    ble_hci_uart_init();
+    nimble_port_init();
+    ble_hs_sched_start();
+
+    // Wait for sync callback
+    while (ble_state != BLE_STATE_ACTIVE) {
+        MICROPY_EVENT_POLL_HOOK
+    }
+
+    return 0;
+}
+
+void mp_bluetooth_disable(void) {
+    ble_state = BLE_STATE_OFF;
+    mp_hal_pin_low(pyb_pin_BT_REG_ON);
+}
+
+bool mp_bluetooth_is_enabled(void) {
+    return ble_state == BLE_STATE_ACTIVE;
+}
+
+void mp_bluetooth_get_addr(uint8_t *addr) {
+    mp_hal_get_mac(MP_HAL_MAC_BDADDR, addr);
+    // TODO: need to convert MSB/LSB?
+}
+
+int mp_bluetooth_advertise_start(bool connectable, uint16_t interval_ms, const uint8_t *adv_data, size_t adv_data_len, const uint8_t *sr_data, size_t sr_data_len) {
+    int ret;
+
+    mp_bluetooth_advertise_stop();
+
+    if (adv_data != NULL) {
+        ret = ble_gap_adv_set_data(adv_data, adv_data_len);
+        if (ret != 0) {
+            //printf("ble_gap_adv_set_data: fail with %u\n", ret);
+            return ble_hs_err_to_errno(ret);
+        }
+    }
+
+    if (sr_data != NULL) {
+        ret = ble_gap_adv_rsp_set_data(sr_data, sr_data_len);
+        if (ret != 0) {
+            //printf("ble_gap_adv_rsp_set_data: fail with %u\n", ret);
+            return ble_hs_err_to_errno(ret);
+        }
+    }
+
+    // Convert from 1ms to 0.625ms units.
+    interval_ms = interval_ms * 8 / 5;
+    if (interval_ms < 0x20 || interval_ms > 0x4000) {
+        return MP_EINVAL;
+    }
+
+    struct ble_gap_adv_params adv_params = {
+        .conn_mode = connectable ? BLE_GAP_CONN_MODE_UND : BLE_GAP_CONN_MODE_NON,
+        .disc_mode = BLE_GAP_DISC_MODE_GEN,
+        .itvl_min = interval_ms,
+        .itvl_max = interval_ms,
+        .channel_map = 7, // all 3 channels
+    };
+
+    ret = ble_gap_adv_start(BLE_OWN_ADDR_PUBLIC, NULL, BLE_HS_FOREVER, &adv_params, gap_event_cb, NULL);
+    if (ret != 0) {
+        //printf("ble_gap_adv_start: fail with %u\n", ret);
+        return ble_hs_err_to_errno(ret);
+    }
+
+    return 0;
+}
+
+void mp_bluetooth_advertise_stop(void) {
+    if (ble_gap_adv_active()) {
+        ble_gap_adv_stop();
+    }
+}
+
+static int characteristic_access_cb(uint16_t conn_handle, uint16_t value_handle, struct ble_gatt_access_ctxt *ctxt, void *arg) {
+    // printf("characteristic_access_cb: conn_handle=%u value_handle=%u op=%u\n", conn_handle, value_handle, ctxt->op);
+    mp_map_elem_t *elem;
+    gatts_db_entry_t *entry;
+    switch (ctxt->op) {
+        case BLE_GATT_ACCESS_OP_READ_CHR:
+            // printf("read_chr\n");
+
+            elem = mp_map_lookup(gatts_db, MP_OBJ_NEW_SMALL_INT(value_handle), MP_MAP_LOOKUP);
+            if (!elem) {
+                return BLE_ATT_ERR_UNLIKELY;
+            }
+            entry = MP_OBJ_TO_PTR(elem->value);
+            os_mbuf_append(ctxt->om, entry->data, entry->data_len);
+
+            return 0;
```

Comment:
It's common on other bluetooth stacks to allow for a callback here when a read request comes in. On NRF devices this is exposed as a callback event BLE_GATTS_EVT_RW_AUTHORIZE_REQUEST.

It's officially used to provide the application with a chance to authorise or deny access to the characteristic based on whether it's paired etc.

It's also commonly used to allow the application to change the value of the characteristic each read. This can be then used to stream data out the characteristic reliably, instead of notifications which are far more unreliable as they don't have the same handshaking to ensure they get through.

The problem in this case is your other events are soft handlers, whereas this would need to be hard / inline to allow the callback to change the return value of this function.